### ISSUE TITLE
VCF failing vcf-validation (vcftools)

### ISSUE THREAD

**Issue by Smeds** [ekg/freebayes#498]
Hi

I have problem with vcf-files, generated by freebayes, and specifically phased variants.

If I run vcf-validator I get the following output:

> column SAMPLE-1 at chr2:113982039 .. FORMAT tag [GL] expected different number of values (expected 3, found 6),FORMAT tag [QA] expected different number of values (expected 1, found 2),FORMAT tag [AO] expected different number of values (expected 1, found 2),FORMAT tag [AD] expected different number of values (expected 2, found 3)
> column SAMPLE-1 at chr2:113982040 .. FORMAT tag [AD] expected different number of values (expected 2, found 3),FORMAT tag [AO] expected different number of values (expected 1, found 2),FORMAT tag [QA] expected different number of values (expected 1, found 2),FORMAT tag [GL] expected different number of values (expected 3, found 6)

These are the lines in the vcf:
> chr2	113982039	rs2863241	A	G	3832.2	PASS	AC=2;AF=1;AN=2;LEN=1;NS=1;TYPE=snp;DECOMPOSED;ANN=G|upstream_gene_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000480684|retained_intron||n.-3985T>C|||||3985|,G|downstream_gene_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000485840|retained_intron||n.*1135T>C|||||1135|,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000429538|protein_coding|10/11|c.1189+2693T>C||||||,G|intron_variant|MODIFIER|PAX8-AS1|ENSG00000189223|transcript|ENST00000456685|antisense|2/5|n.306-1140A>G||||||,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000263335|protein_coding|8/9|c.879+2693T>C||||||,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000348715|protein_coding|10/11|c.1110+2693T>C||||||,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000397647|protein_coding|7/8|c.778-4284T>C||||||,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000465084|protein_coding|2/3|c.207+2637T>C||||||WARNING_TRANSCRIPT_NO_START_CODON,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000497038|nonsense_mediated_decay|4/5|c.*634+2693T>C||||||WARNING_TRANSCRIPT_NO_START_CODON,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000468980|protein_coding|3/4|c.355+2693T>C||||||WARNING_TRANSCRIPT_NO_START_CODON,G|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000263334|protein_coding|9/9|c.1110+2693T>C||||||	GT:GQ:DP:AD:RO:QR:AO:QA:GL	1|1:139:145:0,73,71:0:0:73,71:2461,2416:-394.778,-195.639,-173.664,-199.245,0,-177.872
> chr2	113982040	rs12619508	C	A	3832.2	PASS	AC=1;AF=0.5;AN=2;LEN=1;NS=1;TYPE=snp;DECOMPOSED;ANN=A|upstream_gene_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000480684|retained_intron||n.-3986G>T|||||3986|,A|downstream_gene_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000485840|retained_intron||n.*1134G>T|||||1134|,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000429538|protein_coding|10/11|c.1189+2692G>T||||||,A|intron_variant|MODIFIER|PAX8-AS1|ENSG00000189223|transcript|ENST00000456685|antisense|2/5|n.306-1139C>A||||||,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000263335|protein_coding|8/9|c.879+2692G>T||||||,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000348715|protein_coding|10/11|c.1110+2692G>T||||||,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000397647|protein_coding|7/8|c.778-4285G>T||||||,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000465084|protein_coding|2/3|c.207+2636G>T||||||WARNING_TRANSCRIPT_NO_START_CODON,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000497038|nonsense_mediated_decay|4/5|c.*634+2692G>T||||||WARNING_TRANSCRIPT_NO_START_CODON,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000468980|protein_coding|3/4|c.355+2692G>T||||||WARNING_TRANSCRIPT_NO_START_CODON,A|intron_variant|MODIFIER|PAX8|ENSG00000125618|transcript|ENST00000263334|protein_coding|9/9|c.1110+2692G>T||||||	GT:GQ:DP:AD:RO:QR:AO:QA:GL	1|0:139:145:0,73,71:0:0:73,71:2461,2416:-394.778,-195.639,-173.664,-199.245,0,-177.872

Should AD, GL and QA  look like this?
 * AD = 0,73,71
 * GL = -394.778,-195.639,-173.664,-199.245,0,-177.872
 * QA = 2461,2416

**Comment by ekg** [comment]
How did you produce this file? I don't understand how it came directly out
of freebayes, as the number of GLs does not match the number of alleles.

I see a "decomposed" flag. So this came from a record with more than two
alleles and was broken apart. This is why you're getting the error.

**Comment by ekg** [comment]
No worries! This will maybe help someone else. CC @chapmanb.

**Comment by chapmanb** [comment]
Thanks much for the heads up. Decomposition is definitely imperfect in converting over all of FORMAT variables. It's unfortunately not an easy situation to change. Not decomposing leads to MNPs which don't work as well with downstream tools, and we don't know of a different tool to use that will decompose these correctly. Sorry if it's causing issues in your analysis and hopefully it's possible to work around. Thanks for starting this discussion.

**Comment by tseemann** [comment]
Decomposing MNPs is also the bane of my existence.

**Comment by andreaswallberg** [comment]
I regularly run into the issue of how to pass bialleic SNP GL likelihoods derived from decomposed MNPs to downstream programs, such as BEAGLE.

Here is an example of a decomposed genotype from Freebayes that I find difficult to grasp, which was generated from "vcfallelicprimitives -kg" and is printed in "standard" GT:DP:AD:RO:QR:AO:QA:GL format:

1|0:2:0,0,0,1,0,1:0:0:0,0,1,0,1:0,0,37,0,34:-6.11564,-6.11564,-6.11564,-6.11564,-6.11564,-6.11564,-3.07241,-3.07241,-3.07241,-2.77138,-6.11564,-6.11564,-6.11564,-3.07241,-6.11564,-3.3968,-3.3968,-3.3968,0,-3.3968,-3.09577

As far as I can tell, it is reported as a phased "1|0" genotype but it has 0 observations of the reference allele, and looking at the very long GL chain it seems to have outputted likelihoods for seven different allelic combinations...?

I currently pipe my SNP call through:

`vcfallelicprimitives -kg | vcfsnps | vcfbiallelic`

This produces too many hard to parse genotypes for me to handle. What is the current recommended approach to reduce output of MNPs while still maintaining algorithm power?

Can this be used to avoid "clumping" observations into MNPs?
```
-E --max-complex-gap -1
      --haplotype-length -1
```

Any feedback is much appreciated!
